### PR TITLE
terraform: update all providers and move nixos provider under terraform.withPlugins

### PR DESCRIPTION
**Pull request by kalbasit**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

/cc @jgeerds @zimbatm @peterhoeg

### CODE REVIEW

**Comment by flokli** [comment]
@kalbasit moving "official" and "standalone" providers into the same namespace makes a lot of sense, thanks :-)

Could you also move the two remaining providers, `terraform-provider-ibm` and `terraform-provider-libvirt` in there?

The former one should be completely straightforward, for the latter one we would need to add a way to `overrideAttrs`, setting `(propagated)buildInputs` - either from `update-all`, or in `providers/default.nix`.

**Comment by kalbasit** [comment]
@flokli I tried moving both providers but I hit two issues that I thought were out of scope for this PR:

- `terraform-provider-ibm` provides a go package downloaded from `github.com/IBM-Cloud/terraform-provider-ibm` but the package itself expects to be cloned at `github.com/terraform-providers/terraform-provider-ibm` so this needs a special case in https://github.com/NixOS/nixpkgs/blob/0ad81753c5bea10f4c0bda9afaf76c7952756ab2/pkgs/applications/networking/cluster/terraform/providers/default.nix#L9
- `terraform-provider-libvirt` requires the `cdrtools` `propagatedBuildInputs`.

I'm hoping to first merge this PR, and work on the support for custom attributes in a separate PR.

**Comment by flokli** [comment]
@kalbasit separate PR is fine, too - just wanted to make sure these don't slip through :-)

**Comment by kalbasit** [comment]
@flokli It won't! I'll figure out the best way and submit a new PR sometime this week.

File: pkgs/applications/networking/cluster/terraform/providers/providers.txt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# lines starting with a # are comments
+
+# The accepted format (double escape all grep expresssions):
+#  <organisation>         [grep-expression] [grep-v-expression]  - include all repositories in the organisation.
+#                                                                   grep-expression:   filter repo matching the expression
+#                                                                   grep-v-expression: filter repo not matching the expression
+#  <organisation>/<repo>                                         - include only the named repository.
+
+# include all terraform-providers
+terraform-providers terraform-provider- terraform-provider-azure-classic\\|terraform-provider-scaffolding
```
Comment (Mic92) [inline pkgs/applications/networking/cluster/terraform/providers/providers.txt]: Looks good, but you also need to filter out: terraform-provider-google-beta

```
*** terraform-provider-google-beta ***
fetching https://api.github.com/repos/terraform-providers/terraform-provider-google-beta/git/refs/tags?per_page=100
##O#- #
curl: (22) The requested URL returned error: 404 Not Found
The curl command has failed. Attempt 1/30
```

Comment (kalbasit) [reply pkgs/applications/networking/cluster/terraform/providers/providers.txt]: @Mic92 PTAL.

**Comment by zimbatm** [comment]
@kalbasit just pushed a change to your branch to move the providers to their own `terraform-providers` group. I wanted to add an alias from terraform-provider-nixos and the target wasn't reachable.

For the remaining 2 plugins, since they have special requirements they could just be moved under that group with callPackage links so the generic builder doesn't have to be extended.

**Comment by kalbasit** [comment]
@zimbatm thx for the contribution but the eval is failing, can you take another look at it?

**Comment by kalbasit** [comment]
@zimbatm never mind, I managed to fix it. The issue was that `recurseIntoAttrs` set up the packages as well as the functions `override`, `overrideDerivation` and the `recurseForDerivations` boolean. Removing them fixed the issue.

**Comment by zimbatm** [comment]
nice work @kalbasit

**Comment by zimbatm** [comment]
ported to 18.09 with `git cherry-pick -x -m 1 5547a8b`

**Pull request merged**